File: hello_pwa_with_serviceworker_and_dart/web/sw.dart
import 'package:service_worker/worker.dart' as sw;

void main() {
  String cacheName = "v0.0.1";

  sw.onInstall.listen((sw.InstallEvent e) async {
    print("##>> install");
    Future installTask = new Future(() async {
      sw.Cache cache = await sw.caches.open(cacheName);
      cache.addAll([
        "/","/index.html"
      ]);
    });
    e.waitUntil(installTask);
  });

  sw.onFetch.listen((sw.FetchEvent e) {
    print("##>> fetch");
    Future<sw.Response> responseTask = new Future(() async {
      sw.Response resonse = await sw.caches.match(e.request);
      if(resonse != null){
        return resonse;
      } else {
        return sw.fetch(e.request);
      }
    });
    e.respondWith(responseTask);
  });
}

File: hello_pwa_with_javascript/sw.js
var cacheName = 'hello-v0.0.1';

self.addEventListener('install', function (event) {
  console.log('# install (1)');
  const installFunc = function (cache) {
    console.log('# install (2)');
    return cache.addAll(["/", "/index.html"]);
  };
  event.waitUntil(caches.open(cacheName).then(installFunc));
});

self.addEventListener('fetch', function (event) {
  const responceFunc = function (response) {
    console.log("# fetch : " + event.request.url);
    if (response) {
      return response;
    }
    return fetch(event.request);
  };
  //https://developer.mozilla.org/en/docs/Web/API/FetchEvent/request
  //https://developer.mozilla.org/en/docs/Web/API/FetchEvent/respondWith
  //https://developer.mozilla.org/en/docs/Web/API/Cache/match
  event.respondWith(caches.match(event.request).then(responceFunc));
});

